src/optimizer/join_order/cardinality_estimator.cpp

```diff
@@ -123,44 +123,59 @@ void CardinalityEstimator::RemoveEmptyTotalDomains() {
 	relations_to_tdoms.erase(remove_start, relations_to_tdoms.end());
 }
 
-void UpdateDenom(Subgraph2Denominator &relation_2_denom, RelationsToTDom &relation_to_tdom) {
+void UpdateDenom(Subgraph2Denominator &relation_2_denom, RelationsToTDom &relation_to_tdom, FilterInfo *filter) {
 	relation_2_denom.denom *= relation_to_tdom.has_tdom_hll ? relation_to_tdom.tdom_hll : relation_to_tdom.tdom_no_hll;
 }
 
-void FindSubgraphMatchAndMerge(Subgraph2Denominator &merge_to, idx_t find_me,
+bool FindSubgraphMatchAndMerge(Subgraph2Denominator &merge_to, idx_t find_me,
                                vector<Subgraph2Denominator>::iterator subgraph,
                                vector<Subgraph2Denominator>::iterator end) {
 	for (; subgraph != end; subgraph++) {
 		if (subgraph->relations.count(find_me) >= 1) {
 			for (auto &relation : subgraph->relations) {
 				merge_to.relations.insert(relation);
+				merge_to.numerator_relations.insert(relation);
 			}
+			// remove the relations so this subgraph
+			// can be deleted from the subgraph2Denominator array
 			subgraph->relations.clear();
 			merge_to.denom *= subgraph->denom;
-			return;
+			return true;
 		}
 	}
+	return false;
 }
 
-template <>
-double CardinalityEstimator::EstimateCardinalityWithSet(JoinRelationSet &new_set) {
-
-	if (relation_set_2_cardinality.find(new_set.ToString()) != relation_set_2_cardinality.end()) {
-		return relation_set_2_cardinality[new_set.ToString()].cardinality_before_filters;
+vector<Subgraph2Denominator>::iterator FindMatchingSubGraph(Subgraph2Denominator &merge_to, idx_t find_me,
+                                                            vector<Subgraph2Denominator>::iterator subgraph,
+                                                            vector<Subgraph2Denominator>::iterator end) {
+	for (; subgraph != end; subgraph++) {
+		if (subgraph->relations.count(find_me) >= 1) {
+			return subgraph;
+		}
 	}
-	double numerator = 1;
-	unordered_set<idx_t> actual_set;
+	return subgraph;
+}
 
-	for (idx_t i = 0; i < new_set.count; i++) {
-		auto &single_node_set = set_manager.GetJoinRelation(new_set.relations[i]);
+double CardinalityEstimator::GetNumerator(unordered_set<idx_t> &set) {
+	double numerator = 1;
+	for (auto &relation_id : set) {
+		auto &single_node_set = set_manager.GetJoinRelation(relation_id);
 		auto card_helper = relation_set_2_cardinality[single_node_set.ToString()];
 		numerator *= card_helper.cardinality_before_filters == 0 ? 1 : card_helper.cardinality_before_filters;
-		actual_set.insert(new_set.relations[i]);
 	}
+	return numerator;
+}
 
+DenomInfo CardinalityEstimator::GetDenominator(JoinRelationSet &set) {
```

Comment:
This function (previously `EstimateCardinalityWithSet`, now `GetDenominator`) is quite long (~170 lines), and I find it difficult to understand what is happening. Do you think it's possible to refactor this?

Perhaps we can change some of the `if`/`else` below to this:
```c++
switch(filter->join_type) {
case JoinType::INNER:
    UpdateDenominatorInner(...);
case JoinType::SEMI:
case JoinType::ANTI:
    UpdateDenominatorSemiAnti(...);
default:
    D_ASSERT(filter->join_type == JoinType::INVALID);
    // Not sure if it should be INVALID, I just like assertions
    UpdateDenominatorCrossProduct(...);
}
```

I find using switches and separating logic into different functions improves readability.